### PR TITLE
Implement WorldBorderMock and WorldMock#getWorldBorder

### PR DESCRIPTION
**Pull request by ohowe1**
# Description
State the changes of the pull request here.
Relatively straight forward pull request. Adds `WorldBorderMock` and implements it into `WorldMock`

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Version number updated in `settings.gradle`.
- [x] Unit tests added.
- [x] Code follows existing code format.

# Info on creating a pull request
Before the pull request can be accepted, the version number stored in settings.gradle should be updated.
MockBukkit uses semantic-versioning (https://semver.org/)

 - If the PR fixes a bug in MockBukkit, update the patch version. (if the current version is `0.4.1`, the new version should be `0.4.2`)
 - If the PR adds a new feature to MockBukkit, update minor version. (if the current version is `0.4.1`, the new version should be `0.5.0`)
 - Make sure that unit tests are added which test the relevant changes.
 - Make sure that the changes follow the existing code format.

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

The version number can be found in `settings.gradle`.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java [hunk 1]
```diff
@@ -1158,8 +1161,7 @@ public boolean isGameRule(String rule)
 	@Override
 	public WorldBorder getWorldBorder()
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java]: Wrong return type and missing annotation

```suggestion
	public @NotNull WorldBorderMock getWorldBorder()
```

File: src/test/java/be/seeseemelk/mockbukkit/WorldBorderMockTest.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
+package be.seeseemelk.mockbukkit;
+
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.WorldBorder;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (TheBusyBiscuit) [inline src/test/java/be/seeseemelk/mockbukkit/WorldBorderMockTest.java]: wildcard imports *cries*

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+{
+
+	World world;
+	double size;
+	double damageAmount = 0.2;
+	double damageBuffer = 5.0;
+	int warningDistance = 5;
+	int warningTime = 15;
+	double centerX = 0;
+	double centerZ = 0;
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: Missing access modifiers

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
+package be.seeseemelk.mockbukkit;
+
+import org.apache.commons.lang.Validate;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.WorldBorder;
+import org.bukkit.util.BoundingBox;
+import org.jetbrains.annotations.NotNull;
+
+public class WorldBorderMock implements WorldBorder
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java:18]: Some docs would be nice, no necessity though

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java (hunk 3 above)
Comment (TheBusyBiscuit) [reply src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: Also are these the correct default values here?

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+{
+
+	World world;
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: World should be final.

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	@Override
+	public void setSize(double newSize)
+	{
+		setSize(newSize, 0);
+	}
+
+	@Override
+	public void setSize(double newSize, long seconds)
+	{
+		// I dont think that implementing seconds here is very important
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: why.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Not a full review but I spotted some things that should be addressed before continuing.

Thanks for the pull request though!

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java (hunk 3 above)
Comment (ohowe1) [reply src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: > Also are these the correct default values here?

Yep

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+	public WorldBorderMock(World world, Server server)
+	{
+		this.world = world;
+		this.server = server;
+	}
+
+	@Override
+	public void reset()
+	{
+		setSize(6.0E7);
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: `6.0E7` is used twice here, I think there should be a constant for this at the top of the class.
`private static final double DEFAULT_BORDER_SIZE = 6.0E7`

Avoiding magic numbers like these makes it easier to tweak them if this default size changes in future versions of Minecraft.

*Same theoretically goes for the other values.*

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	public void setCenter(@NotNull Location location)
+	{
+		setCenter(location.getX(), location.getZ());
+	}
+
+	@Override
+	public void setCenter(double x, double z)
+	{
+		this.centerX = x;
+		this.centerZ = z;
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: CraftBukkit does a minmax check here before setting these values:
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/CraftWorldBorder.java#63-64

File: src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java [hunk 9]
```diff
@@ -23,7 +10,10 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-import be.seeseemelk.mockbukkit.block.BlockMock;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (TheBusyBiscuit) [inline src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java]: Wildcard imports D:

File: src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java [hunk 10]
```diff
@@ -1,20 +1,7 @@
...
-
-import java.util.List;
-
-import org.bukkit.Chunk;
-import org.bukkit.GameRule;
-import org.bukkit.Location;
-import org.bukkit.Material;
-import org.bukkit.World;
+import be.seeseemelk.mockbukkit.block.BlockMock;
+import org.bukkit.*;
```
Comment (TheBusyBiscuit) [inline src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java]: Do not use a wildcard import here :monkaS:

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	private double centerX;
+	private double centerZ;
+
+	/**
+	 * Creates a new world border mock
+	 *
+	 * @param world The world it is the border of
+	 * @param server The server it is in
+	 */
+	public WorldBorderMock(World world, Server server)
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]:
```suggestion
	public WorldBorderMock(@NotNull World world,@NotNull Server server)
```

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+{
+
+	private static final double DEFAULT_BORDER_SIZE = 6.0E7;
+	private static final double DEFAULT_DAMAGE_AMOUNT = 0.2;
+	private static final double DEFAULT_DAMAGE_BUFFER = 5.0;
+	private static final int DEFAULT_WARNING_DISTANCE = 5;
+	private static final int DEFAULT_WARNING_TIME = 15;
+	private static final double DEFAULT_CENTER_X = 0;
+	private static final double DEFAULT_CENTER_Z = 0;
+	private static final double MAX_CENTER_VALUE = 3.0E7D;
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: Why is `MAX_CENTER_VALUE` the only one with a type literal suffix?

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+					{
+						size += distancePerTick;
+					}
+					else
+					{
+						size = newSize;
+						this.cancel();
+					}
+				}
+			}, 1, 1);
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: We can use a lambda-expression here.
Also is `null` valid to use here?

```suggestion
			server.getScheduler().runTaskTimer(null, () ->
			{
				if ((size < newSize && distance > 0) || (size > newSize && distance < 0))
				{
					size += distancePerTick;
				}
				else
				{
					size = newSize;
					this.cancel();
				}
			}, 1, 1);
```

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+		}
+		else
+		{
+			setSize(newSize);
+		}
+	}
+
+	@NotNull
+	@Override
+	public Location getCenter()
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]:
```suggestion
	@Override
	public @NotNull Location getCenter()
```

**Review by TheBusyBiscuit — DISMISSED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java (hunk 13 above)
Comment (ohowe1) [reply src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: > Also is `null` valid to use here?

Unless there is an system for internally scheduling repeating tasks inside mockbukkit, then I think that this is the only way to do it. The null should not be a problem as all plugin is used for is canceling tasks that were created by the plugin in `BukkitSchedulerMock#cancelTasks`.

Comment (ohowe1) [reply src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: > We can use a lambda-expression here.

It appears that a lambada cannot be used as the cancel method is called.

**Comment by seeseemelk** [comment]
@TheBusyBiscuit Are you still requesting changes here?

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+			double ticksToTake = seconds * 20;
+
+			double distancePerTick = distance / ticksToTake;
+
+			server.getScheduler().runTaskTimer(null, new BukkitRunnable()
+			{
+				@Override
+				public void run()
+				{
+					if ((size < newSize && distance > 0) || (size > newSize && distance < 0))
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: Floating point values may not always converge to exactly 0.0

**Review by TheBusyBiscuit — COMMENTED** [review]
Sorry for the late response!

Most of my previous requests have been addressed.
Only things I am pondering about are param validation and floating point issues.
Other than that, it looks mergeable @seeseemelk

File: src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java (hunk 15 above)
Comment (ohowe1) [reply src/main/java/be/seeseemelk/mockbukkit/WorldBorderMock.java]: I think you are referring to the size comparison to newSize, correct? So if I'm not completely misunderstanding then something like `newSize - size < delta` would work? As in if size is bigger than newSize within this margin.

**Review by Insprill — APPROVED** [review]

**Pull request merged**